### PR TITLE
New License: OPL-UK-3.0

### PR DESCRIPTION
**Pull request by BanulaKumarage**
Test - (but could potentially fix https://github.com/spdx/license-list-XML/issues/1788 )

### CODE REVIEW

**Comment by jlovejoy** [comment]
Hi @BanulaKumarage - for context, did you create this XML file by using the online tool via https://tools.spdx.org/app/license_requests/203/ ?

**Comment by BanulaKumarage** [comment]
@jlovejoy yes I created using the online tool. I'll create the .txt file as well.

**Comment by BanulaKumarage** [comment]
@jlovejoy I updated the PR.

**Comment by jlovejoy** [comment]
Hi @BanulaKumarage - okay, so I ran the tests and some are failing. But this is a good exercise for you to see if you spot the issues, here's some thoughts on what to look at:

When you used the online tool to create the xml file - did you edit the xml at all or just take the xml and submit as it was?
If you submitted "as is", then please review https://github.com/spdx/license-list-XML/blob/main/DOCS/xml-fields.md for some hints on some of the edits that the xml file needs. I can see a few things that need to be modified in the xml file. This is also "testing" our documentation, which are you also welcome to provide input on - whether you have questions or if you have ideas for improvements. It's all part of the process! Our documentation for this process needs to be easy enough to follow for someone who is not a "coder" or who may have never used this tool.

In regard to some of the manual edits you might need to make to the xml file - you might consider if any of that can be added to the functionality of the tool/be automated. (I don't think much more can be automated, but you may have a different perspective!)

File: src/United.xml [hunk 1]
```diff
@@ -0,0 +1,168 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<SPDXLicenseCollection xmlns="http://www.spdx.org/license">
+   <license isOsiApproved="false" licenseId="OPL-UK-3.0" name="United
+   Kingdom Open Parliament Licence v3.0" listVersionAdded="">
+      <crossRefs>
+         <crossRef>https://www.parliament.uk/site-information/copyright-parliament/open-parliament-licence/</crossRef>
+      </crossRefs>
+      <standardLicenseHeader>
+      </standardLicenseHeader>
+      <notes>
```
Comment (jlovejoy) [inline src/United.xml:10]: a couple hints: the titleText tag is missing and the standardLicenseHeader tag is not needed and can be removed.
you can also probably remove the notes tag, as there are none (that we know of, at this point)

**Comment by jlovejoy** [comment]
also note... filenames :)

**Comment by BanulaKumarage** [comment]
> Hi @BanulaKumarage - okay, so I ran the tests and some are failing. But this is a good exercise for you to see if you spot the issues, here's some thoughts on what to look at:
> 
> When you used the online tool to create the xml file - did you edit the xml at all or just take the xml and submit as it was? If you submitted "as is", then please review https://github.com/spdx/license-list-XML/blob/main/DOCS/xml-fields.md for some hints on some of the edits that the xml file needs. I can see a few things that need to be modified in the xml file. This is also "testing" our documentation, which are you also welcome to provide input on - whether you have questions or if you have ideas for improvements. It's all part of the process! Our documentation for this process needs to be easy enough to follow for someone who is not a "coder" or who may have never used this tool.
> 
> In regard to some of the manual edits you might need to make to the xml file - you might consider if any of that can be added to the functionality of the tool/be automated. (I don't think much more can be automated, but you may have a different perspective!)

Hi @jlovejoy I haven't edited the xml. I submitted it as well. I will look into the [xml fileds](https://github.com/spdx/license-list-XML/blob/main/DOCS/xml-fields.md) and try to edit the xml file. It will be a good learning for me as candidate for GSOC.

File: src/United.xml (hunk 1 above)
Comment (BanulaKumarage) [reply src/United.xml:10]: @jlovejoy I will look into those issues and get back to you if any clarification needed.

**Comment by BanulaKumarage** [comment]
> also note... filenames :)

If I use SGP4-Unlicense.xml and SGP4-Unlicense.txt as file names, will it be fine?

**Comment by BanulaKumarage** [comment]
@jlovejoy Got the following error I validated using the tool. 
image

Can you briefly explain the error?

**Comment by goneall** [comment]
@BanulaKumarage - It is failing validation against the [XML schema](https://github.com/spdx/license-list-XML/blob/main/schema/ListedLicense.xsd).  There was a recent change to the schema in this area to match the HTML validation.

Looking at the XML file, on line 69 there is a `<p>` tag which is not allowed.  You can only have `<item>`'s - you can put `<p>` tags inside of items.  From a quick glance, it looks like the previous list should be ended, then the paragraph, then a new list started.

**Comment by BanulaKumarage** [comment]
> @BanulaKumarage - It is failing validation against the [XML schema](https://github.com/spdx/license-list-XML/blob/main/schema/ListedLicense.xsd). There was a recent change to the schema in this area to match the HTML validation.
> 
> Looking at the XML file, on line 69 there is a `<p>` tag which is not allowed. You can only have `<item>`'s - you can put `<p>` tags inside of items. From a quick glance, it looks like the previous list should be ended, then the paragraph, then a new list started.
Thanks @goneall. Got the idea

**Comment by BanulaKumarage** [comment]
@jlovejoy earlier the xml is failed in the validation in online tools validation option. I have edited and now the validation is passed. And also added your suggestions as well. Please run the tests.

**Comment by BanulaKumarage** [comment]
Hi @jlovejoy, can you explain what refers to canonical match?

File: src/OPL-UK-3.0.xml [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+         <list>
+            <item>
+               <bullet>*</bullet>
+               <p>
+                  copy, publish, distribute and transmit the information
+                  adapt the information exploit the information
+                  commercially and non-commercially, for example,
+                  by combining it with other information, or by
+                  including it in your own product or application
+               </p>
```
Comment (jlovejoy) [inline src/OPL-UK-3.0.xml:47]: looking at how the license displays in its original form, there should be 3 list items (i.e., the lines starting with "copy...", "adapt..", and "exploit..." as you see https://www.parliament.uk/site-information/copyright-parliament/open-parliament-licence/

File: src/OPL-UK-3.0.xml [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+               <p>
+                  You must (where you do any of the above):
+               </p>
```
Comment (jlovejoy) [inline src/OPL-UK-3.0.xml:50]: the above list should then end and a new one begin (this is where the XML editor doesn't always "read" lists and nested lists so well and some human editing is needed)

File: src/OPL-UK-3.0.xml [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+                  attribution statement and, where possible, provide a
+                  link to this licence: Contains Parliamentary information
+                  licensed under the Open Parliament Licence v3.0.
+               </p>
+               <p>
+                  These are important conditions of this licence and
+                  if you fail to comply with them the rights granted to
+                  you under this licence, or any similar licence granted
+                  by the Licensor, will end automatically. Exemptions
+               </p>
```
Comment (jlovejoy) [inline src/OPL-UK-3.0.xml:66]: Exemptions should be on it's own line and same issue as above for the various list items

**Comment by jlovejoy** [comment]
@BanulaKumarage - I'm not sure why it's still failing (I'm not so good at reading the errors, so I usually just try to look/compare the xml and txt files), but I put some comments in regarding fixing the formatting around the list items in the xml file which should correspond to how the bullets display on the website for the license (hopefully that makes sense)

This may also need some changes to the text file.

**Comment by BanulaKumarage** [comment]
@jlovejoy is there a way to run those two tests from my end without updating the PR?

**Comment by jlovejoy** [comment]
@BanulaKumarage - this is now an accepted license, so let's see if we can get this sorted out and merged!
I'm not sure if you addressed the above formatting issues with the bullets?
also need to add 3.21 for the license list version (assuming we can get things fixed quickly)

**Comment by BanulaKumarage** [comment]
> @BanulaKumarage - this is now an accepted license, so let's see if we can get this sorted out and merged! I'm not sure if you addressed the above formatting issues with the bullets? also need to add 3.21 for the license list version (assuming we can get things fixed quickly)

Some of the formatting have to be done. I will look into those. Lets get this merged

**Comment by BanulaKumarage** [comment]
@jlovejoy the files have been updated according to the  [ https://www.parliament.uk/site-information/copyright-parliament/open-parliament-licence/](https://www.parliament.uk/site-information/copyright-parliament/open-parliament-licence/).

Can you run the tests again?

**Comment by jlovejoy** [comment]
will do, but I think the licenseListVersion (3.21) is still missing :)

**Comment by BanulaKumarage** [comment]
> will do, but I think the licenseListVersion (3.21) is still missing :)

Can you please show the place where should I add the licenseListVersion?


@jlovejoy in the PR build in the canonical match test it shows as below.
image

Can you explain this issue?

**Comment by BanulaKumarage** [comment]
@jlovejoy haven't changed the file. No need of running the tests

**Comment by jlovejoy** [comment]
@BanulaKumarage - looks like you made a commit to add the license list version and then remove it?  I'm going to just fix that as it's not worth going back and forth on that!

then I'll look closer overall

**Comment by BanulaKumarage** [comment]
> @BanulaKumarage - looks like you made a commit to add the license list version and then remove it? I'm going to just fix that as it's not worth going back and forth on that!
> 
> then I'll look closer overall

@jlovejoy I removed it since the field name is `listVersionAdded` but not licenseListVersion. Are both the same?

**Comment by BanulaKumarage** [comment]
@jlovejoy the test failed with 
```
The following warning(s) were identified:
	Test for license OPL-UK-3.0 failed: Test for license ID OPL-UK-3.0 failed due to difference found Normal text of license does not match starting at line #10 column #45 ")" when comparing to template text "

Open Parliament Licence
```
Can you please explain this issue?

**Comment by jlovejoy** [comment]
> > @BanulaKumarage - looks like you made a commit to add the license list version and then remove it? I'm going to just fix that as it's not worth going back and forth on that!
> > then I'll look closer overall
> 
> @jlovejoy I removed it since the field name is `listVersionAdded` but not licenseListVersion. Are both the same?

sorry, my bad, I typed the tag wrong above. better to refer to https://github.com/spdx/license-list-XML/blob/main/DOCS/xml-fields.md than type tags off the top of my head!

**Comment by BanulaKumarage** [comment]
> > > @BanulaKumarage - looks like you made a commit to add the license list version and then remove it? I'm going to just fix that as it's not worth going back and forth on that!
> > > then I'll look closer overall
> > 
> > 
> > @jlovejoy I removed it since the field name is `listVersionAdded` but not licenseListVersion. Are both the same?
> 
> sorry, my bad, I typed the tag wrong above. better to refer to https://github.com/spdx/license-list-XML/blob/main/DOCS/xml-fields.md than type tags off the top of my head!

Ah got it. Thanks @jlovejoy

File: test/simpleTestForGenerator/OPL-UK-3.0.txt [hunk 5]
```diff
@@ -0,0 +1,113 @@
+United Kingdom Open Parliament Licence v3.0
+
+Open Parliament Licence
+
+You are encouraged to use and re-use the information that
+is available under this licence freely and flexibly, with
+only a few conditions. Using information under this licence
+
+Use of copyright and database right material made
+available under this licence (the information) indicates
```
Comment (jlovejoy) [inline test/simpleTestForGenerator/OPL-UK-3.0.txt:10]: aha! I found the problem - (the information) needs to be (the 'information') to match!

**Review by jlovejoy — APPROVED** [review]

**Comment by BanulaKumarage** [comment]
@jlovejoy is there a way to run those tests locally?

**Comment by jlovejoy** [comment]
@BanulaKumarage - there is a way to run the tests locally, but I never do that and am not the one to explain how! (I think you are also experiencing first-hand the challenge of reading the failure notifications though! I struggle with figuring out what they mean as well...)

@swinslow - do we have anything documented anywhere as to how to run the tests locally?

**Comment by BanulaKumarage** [comment]
> @BanulaKumarage - there is a way to run the tests locally, but I never do that and am not the one to explain how! (I think you are also experiencing first-hand the challenge of reading the failure notifications though! I struggle with figuring out what they mean as well...)
> 
> @swinslow - do we have anything documented anywhere as to how to run the tests locally?

Yes @jlovejoy. I am trying to figure out why the validate canonical match fails. Do u have any idea regarding it?

image

**Comment by jlovejoy** [comment]
I'm at a loss here too - the first item that says "Unable to find the text" - I see that text in both files.
And I can't see anything obviously wrong at line 115 column 50 (which I presume is referring to the xml file)

Let's wait until maybe @swinslow or @goneall or @zvr maybe has a moment to have a look. 

Sorry I am not more help here!

**Comment by BanulaKumarage** [comment]
> I'm at a loss here too - the first item that says "Unable to find the text" - I see that text in both files.
> And I can't see anything obviously wrong at line 115 column 50 (which I presume is referring to the xml file)
> 
> Let's wait until maybe @swinslow or @goneall or @zvr maybe has a moment to have a look. 
> 
> Sorry I am not more help here!

Its okay @jlovejoy. Will wait for their reply.

**Comment by zvr** [comment]
@BanulaKumarage it seems our schema does not allow `<p>`s directly after `<bullet>`s.

Delete the `<p>`/`</p>` in lines 49/53, 62/68, 89/94, etc. and leave the text bare following the bullet (like in lines 41, 45, 85, etc.) and it should be ok.

**Comment by BanulaKumarage** [comment]
> @BanulaKumarage it seems our schema does not allow `<p>`s directly after `<bullet>`s.
> 
> Delete the `<p>`/`</p>` in lines 49/53, 62/68, 89/94, etc. and leave the text bare following the bullet (like in lines 41, 45, 85, etc.) and it should be ok.

Done the updates. @jlovejoy please run the tests.

**Comment by jlovejoy** [comment]
Thanks @zvr ! I always get confused on that one... probably something we should add to https://github.com/spdx/license-list-XML/blob/main/DOCS/xml-fields.md

**Comment by BanulaKumarage** [comment]
> Thanks @zvr ! I always get confused on that one... probably something we should add to https://github.com/spdx/license-list-XML/blob/main/DOCS/xml-fields.md
> 
> 

@jlovejoy just an idea. If we can intergrate this canonical validation to our online tools like xml validator, we can validate first and then make the PR.

**Comment by jlovejoy** [comment]
@BanulaKumarage - I'm not sure what that means... 


@zvr do you have any further thoughts on this? If not, we may need @goneall to take a look when he's back online

**Comment by zvr** [comment]
Line 100 still has a `<p>` right after a `<bullet>` ...

**Comment by jlovejoy** [comment]
@goneall - do you think this is the same issue as I was having with #1913 ?

**Comment by jlovejoy** [comment]
created #2008 to see if I could get it to pass. closing this one now

**Pull request closed without merging**